misk-testing/src/main/kotlin/misk/containers/Containers.kt

```diff
@@ -66,154 +61,19 @@ data class Container(
-      require(create.image != "todo") {
-        "must provide an image for container ${create.name}"
-      }
-
-      docker.listContainersCmd()
-        .withShowAll(true)
-        .withLabelFilter(mapOf("name" to name))
-        .exec()
-        .forEach {
-          log.info { "removing previous $name container with id ${it.id}" }
-          docker.removeContainerCmd(it.id).exec()
-        }
-
-      log.info { "pulling ${create.image} for $name container" }
-
-      val imageParts = create.image!!.split(":")
-      docker.pullImageCmd(imageParts[0])
-        .withTag(imageParts.getOrElse(1) { "latest" })
-        .exec(PullImageResultCallback()).awaitCompletion()
-
-      log.info { "starting $name container" }
-
-      val id = create
-        .withNetworkMode(network.id())
-        .withLabels(mapOf("name" to name))
-        .withTty(true)
-        .exec()
-        .id
-      containerIds[name] = id
-
-      container.beforeStartHook(docker, id)
-
-      docker.startContainerCmd(id).exec()
-      docker.logContainerCmd(id)
-        .withStdErr(true)
-        .withStdOut(true)
-        .withFollowStream(true)
-        .withSince(0)
-        .exec(LogContainerResultCallback())
-        .awaitStarted()
-
-      log.info { "started $name; container id=$id" }
-    }
-  }
-
-  private fun Container.name(): String {
-    val create = docker.createContainerCmd("todo").apply(createCmd)
-    require(!create.name.isNullOrBlank()) {
-      "must provide a name for the container"
-    }
-    return "$name/${create.name}"
-  }
-
-  fun stop() {
-    if (!running.compareAndSet(true, false)) return
-
-    for (container in containers) {
-      val name = container.name()
-      val id = containerIds[name]!!
-      log.info { "killing $name with container id $id" }
-      docker.removeContainerCmd(id).withForce(true).exec()
-
-      try {
-        log.info { "waiting for $name to terminate" }
-        docker.waitContainerCmd(id).exec(
-          GracefulWaitContainerResultCallback()
-        ).awaitCompletion()
-      } catch (th: Throwable) {
-        log.error(th) { "could not kill $name with container id $id" }
-      }
-
-      log.info { "killed $name with container id $id" }
-    }
-
-    network.stop()
-  }
-
-  private class LogContainerResultCallback :
-    ResultCallbackTemplate<LogContainerResultCallback, Frame>() {
-    override fun onNext(item: Frame) {
-      String(item.payload).trim().split('\r', '\n').filter { it.isNotBlank() }.forEach {
-        log.info(it)
-      }
-    }
-  }
-
-  private class GracefulWaitContainerResultCallback : WaitContainerResultCallback() {
-    override fun onError(throwable: Throwable?) {
-      // this is ok, just meant that the container already terminated before we tried to wait
-      if (throwable is NotFoundException) {
-        return
-      }
-      super.onError(throwable)
-    }
-  }
+  var delegate: wisp.containers.Composer
```

Comment:
nit: Something similar to this so `delegate` is a `val`? (Also, `private/internal val`?)
```kotlin
val delegate: Composer = run {
  // ... init code
}
```
?

Reply:
Yeah, should have changed that - will fix.